invoices+channeldb: reject payments to expired invoices

bhandras opened the pull request:
invoices+channeldb: reject payments to expired invoices

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lightningnetwork/lnd/3694)
<!-- Reviewable:end -->

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -1170,6 +1170,7 @@ func copyInvoice(src *Invoice) *Invoice {
 		Receipt:        copySlice(src.Receipt),
 		PaymentRequest: copySlice(src.PaymentRequest),
 		FinalCltvDelta: src.FinalCltvDelta,
+		Expiry:         src.Expiry,
```
Commit doesn't adhere to the contribution guidelines section about commit message crafting: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#45-ideal-git-commit-structure

bhandras replied:
done

Roasbeef commented on invoices/invoiceregistry.go:
```diff
@@ -4,6 +4,7 @@ import (
 	"errors"
 	"sync"
 	"sync/atomic"
+	"time"
```
Similar comment here re commit message structure (line folding, message header, etc).

bhandras replied:
done

Roasbeef commented on invoices/invoiceregistry.go:
```diff
@@ -12,6 +13,10 @@ import (
 )
 
+const (
+	defaultInvoiceExpiry = 3600 * time.Second
```
Missing a `godoc` comment.

joostjager replied:
This constant is already defined in `zpay32.Invoice.Expiry()`

bhandras replied:
Moved constant to `channeldb/invoices.go` instead as the code has changed during the fixup. IMHO it's not worth using `zpay32.Invoice.Expiry()` here as that'd require creating a zpay32.Invoice.

joostjager replied:
You can extract `3600 * time.Second` into an exported symbol in `zpay32`

Roasbeef commented on invoices/invoiceregistry.go:
```diff
@@ -491,6 +496,18 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
+			invExpiry = defaultInvoiceExpiry
+		}
+
+		// Check invoice expiry.
```
Comments in this diff are borderline compliant with the comments section: https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#43-code-documentation-and-commenting 

> Comments in the body of the code are highly encouraged, but they should explain the intention of the code as opposed to just calling out the obvious

bhandras replied:
PTAL

Roasbeef commented on invoices/invoiceregistry_test.go:
```diff
@@ -240,6 +241,69 @@ func TestSettleInvoice(t *testing.T) {
 }
 
+// Tests that invoice is rejected after expiration.
+func TestInvoiceExpiry(t *testing.T) {
```
Can use `t.Parallel()` here, since the test doesn't interact with any global state.

bhandras replied:
done

Roasbeef commented on invoices/invoiceregistry_test.go:
```diff
@@ -240,6 +241,69 @@ func TestSettleInvoice(t *testing.T) {
+	hodlChan := make(chan interface{}, 1)
+
+	// Sleep just enough to expire the invoice.
+	time.Sleep(testExpiry + time.Millisecond)
```
Alternatively, perhaps we can just make an invoice that has already expired? So craft the creation date and expiry to already have expired by the time the invoice has been created.

Roasbeef commented on invoices/invoiceregistry_test.go:
```diff
@@ -240,6 +241,69 @@ func TestSettleInvoice(t *testing.T) {
+		testHtlcExpiry, testCurrentHeight,
+		getCircuitKey(1), hodlChan, nil,
+	)
+
```
Style nit: extraneous newline.

bhandras replied:
PTAL

joostjager commented on invoices/invoiceregistry_test.go:
```diff
@@ -240,6 +241,69 @@ func TestSettleInvoice(t *testing.T) {
+			PaymentPreimage: preimage,
+			Value:           lnwire.MilliSatoshi(100000),
+		},
+		CreationDate: time.Now(),
```
Better to use a fixed time for testing, to make it deterministic.

bhandras replied:
Sidenote: There are two tests now. One tests AddInvoice with an expired timestamp (one day in the past). The other checking expration with a valid timestamp.

re fixed time: I'm a bit unsure how I can make this completely fixed as expiration checking uses time.Now() as a reference point. Is there maybe a way to mock time in tests?

joostjager replied:
We solved this problem elsewhere by adding a field `now func() time.Time` to `InvoiceRegistry`. That way the test can control the time.

bhandras replied:
done

joostjager commented on invoices/invoiceregistry_test.go:
```diff
@@ -240,6 +241,69 @@ func TestSettleInvoice(t *testing.T) {
+		}
+	case <-time.After(testTimeout):
+		t.Fatal("no update received")
+	}
```
This assert doesn't seem to add much for this test. A new invoice notification is covered already elsewhere.

joostjager replied:
Better would it be to assert that the invoice moves into the `Canceled` state

bhandras replied:
PTAL

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -491,6 +496,18 @@ func (i *InvoiceRegistry) NotifyExitHopHtlc(rHash lntypes.Hash,
+		if inv.CreationDate.Add(invExpiry).Before(time.Now()) {
+			debugLog("invoice expired")
+			return nil, errNoUpdate
+		}
```
Main question about this PR is whether we should move the invoice to the `Canceled` state here.

bhandras replied:
PTAL. I changed the logic such that:
1) AddInvoice does not accept expired invoices anymore (even though there are no code paths currently doing this).
2) NotifyExpitHopHtlc first checks if the invoice has expired and cancels it if that is the case, otherwise there is no change in existing behavior.

joostjager commented:
Working on a different timer in https://github.com/lightningnetwork/lnd/pull/3415 using a priority queue

alexbosworth commented:
Can we change the default invoice time to be higher than an hour? How about 3 hours? Otherwise I anticipate this change resulting in some error rate increases.

joostjager commented:
Why would error rates increase? The sender already checks the expiry now. Or are you thinking about other implementations that may not do this? Or people decoding the pay req and using `SendPayment`?

alexbosworth commented:
> Why would error rates increase? The sender already checks the expiry now. Or are you thinking about other implementations that may not do this? Or people decoding the pay req and using `SendPayment`?

There is decoding and sending yes, but there is also a natural race condition between acceptance and sending. It can easily take me 10 minutes to complete a payment to some destinations, but the timer would have shown "you have 10 minutes left" when I pressed the button. Clocks may not be perfectly synchronized, etc.

joostjager commented:
What does changing the default fix about this? The problem remains, only with a different time.

alexbosworth commented:
I didn't say it fixed it, it just can counterbalance it

halseth commented:
> Perhaps not in this PR, but eventually we also may want to actively cancel invoices when they expire. So that there state is reflectly properly even without someone paying to them first.

I would prefer this being a follow up. Can also make an issue on it for someone to pick up.

joostjager commented:
>I would prefer this being a follow up. Can also make an issue on it for someone to pick up.

@bhandras and I discussed this offline. Both approaches aren't compatible with each other, so doesn't work very well as a follow up.

bhandras commented:
Even though the original issue is solved with the current PR, as discussed with @joostjager, it is cleaner to implement a watcher for invoice expiry and cancellation. I'll update this PR soon.

alexbosworth commented:
Here is a data point on how many errors are due to expiry

- 708 - payment not completed before timeout
- 603 - invoice already paid
- 392 - invoice expired
- 146 - payment is in transition
- 72 - unable to find a path

https://twitter.com/pwkad/status/1196124533759062019

bhandras commented:
I've done some additional work to handle expirations actively as suggested by @joostjager. There's a bit of work left, but feel free to take a look at this. I feel like this thing deserves its own PR now as it is a complete rewrite based on a different idea than the original fix. WDYT?

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,160 @@
+	paymentHash, err := ew.decodePaymentHash(invoice)
+	if err != nil {
+		return fmt.Errorf("Could not decode payment hash for invoice: %v", invoice)
+	}
```
Discussed offline. Data is already present in database. Perhaps only add the `PaymentHash` field to the in memory `Invoice` and populate it in `invoices.go`

joostjager replied:
One other reason to get rid of the payment request decoding is that we won't (naturally) have payment requests for spontaneous payments that we want to implement in the future.

bhandras replied:
This is now done differently as requested.

bhandras commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,160 @@
+// Start starts the expry watcher queue and the subscription handler loop of the
+// InvoiceExpiryWatcher instance. Start() can only be called once.
+func (ew *InvoiceExpiryWatcher) Start() {
+	ew.started.Do(func() {
```
sync.Once for starting/stopping is used quite a few places actually.

joostjager replied:
But is it good? Imo it can cover up a bug.

bhandras replied:
Would be nicer if it'd optionally panic on subsequent calls. Decied to remove sync.Once :)

bhandras commented on channeldb/invoices.go:
```diff
@@ -94,6 +94,9 @@ const (
 	MaxPaymentRequestSize = 4096
 
+	// DefaultInvoiceExpiry is the default invoice expiry of one hour.
+	DefaultInvoiceExpiry = time.Hour
```
This is now removed (not needed).

bhandras commented on channeldb/invoices.go:
```diff
@@ -307,6 +310,21 @@ type InvoiceUpdateDesc struct {
+	}
+
+	// Check and return if expiry time has passed.
+	return i.CreationDate.Add(expiry).Before(time.Now())
```
Also removed.

bhandras commented on expirywatcher/expiry_queue.go:
```diff
@@ -0,0 +1,55 @@
+	// expiry is the expiry time of the item, used as the priority in the queue.
+	expiry time.Time
+	// index is needed and maintained by heap.Interface methods.
+	index int
```
Not needed, removed. Also the implementation changed slightly and this is now moved to queue/priority_queue.go which is a bit different and more general.

bhandras commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,160 @@
+		// watcher.
+		ew.subscription = ew.registry.SubscribeNotifications(0, 0)
+		ew.wg.Add(1)
+		go ew.subscriptionHandler()
```
Thanks for the suggestion, refactored the code to be much simpler and now the extra goroutine is not needed.

bhandras commented on channeldb/invoices.go:
```diff
@@ -307,6 +310,21 @@ type InvoiceUpdateDesc struct {
 
+// IsExpired checks if an Invoice has expired by checking if the expiration,
+// which is Invoice.CreationDate + Invoice.Expiry has passed.
+func (i *Invoice) IsExpired() bool {
```
Removed.

joostjager commented on channeldb/invoices.go:
```diff
@@ -505,6 +505,61 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+// only unsettled invoices and their payment hashes will be returned, skipping
+// all invoices that are fully settled.
+func (d *DB) FetchAllInvoicesWithPaymentHash(pendingOnly bool) (
+	[]Invoice, []lntypes.Hash, error) {
```
Add `PaymentHash` to `Invoice`? (in-memory field only, not serialized to db) and unify with existing FetchAll method?

bhandras replied:
Done, though imho, this may be problematic. Changes semantics as not all ChannelDB functions can full the PaymentRequest and some users may find this confusing.

bhandras replied:
So after taking a turn to that solution, the current cleaned up commit structure only has `FetchAllInvoicesWithPaymentHashes` and returns an array of structs instread of a tuple of arrays as discussed offline.

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,178 @@
+	}
+
+	expiry := invoice.CreationDate.Add(
+		zpay32.DefaultInvoiceExpiry(invoice.Expiry))
```
Name `DefaultInvoiceExpiry` doesn't fully cover what the function does

bhandras replied:
PTAL

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,178 @@
+		ew.cancelExpiredInvoices()
+
+		select {
+		case <-time.After(ew.nextExpiry()):
```
A `nil` channel is also blocked upon indefinitely. Maybe better than using maxint

bhandras replied:
Done, ptal

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -99,21 +102,26 @@ func NewRegistry(cdb *channeldb.DB, finalCltvRejectDelta int32) *InvoiceRegistry
+	i.expiryWatcher.Start(i.CancelInvoice)
+	i.expiryWatcher.PrefetchInvoices(i.cdb)
 
+	i.wg.Add(1)
```
If this is adding the watcher to the wait group, it is probably better to create a new wait group inside the watcher and wait for it on `Stop`

bhandras replied:
Maybe I misunderstand your comment but, `InvoiceExpiryWatcher` has its own wait group, and waits on it in `InvoiceExpiryWarcher.Stop()`.

joostjager replied:
Ok, this `wg.Add` is just a move, but Github showed it a bit strange.

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,178 @@
+			continue
+
+		case newInvoiceExpiry := <-ew.newInvoices:
+			ew.expiryQueue.Push(newInvoiceExpiry)
```
I still wouldn't mind having those two events inside the main invoice registry main loop to reduce concurrency and any risks associated with that. But like this is ok too.

bhandras replied:
I'd prefer keeping this separate as otherwise we introduce unrelated functionality to the subscriber loop.

bhandras commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,160 @@
+		// Subscribe to all invoice notificatons and start a goroutine to
+		// handle the subscription and forward all new invoices to the expiry
+		// watcher.
+		ew.subscription = ew.registry.SubscribeNotifications(0, 0)
```
Discussed later.

bhandras commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,160 @@
+	case invoice := <-ew.subscription.NewInvoices:
+		// Only care about invoices that are not canceled or settled already.
+		if invoice.Terms.State != channeldb.ContractCanceled &&
+			invoice.Terms.State != channeldb.ContractSettled {
```
Done.

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -184,6 +186,10 @@ type Invoice struct {
 
+	// PaymentHash is an optional field where a payment hash of this invoice
+	// can be stored.
+	PaymentHash []byte
```
This is only _sometimes_ populated when things are read into memory? Not sure if we needed to add this field given that we already have an index over this in the database, and it's available as is within the struct, albeit indirectly.

bhandras replied:
This is only populated in FetchAllInvoices. I myself don't like this solution and only did it on suggestion by other reviewers. The only other way of retrieving the payment hash as far as I can see is by decoding the payment request (which was implemented in an earlier incarnation of this PR).

bhandras replied:
Discussed this further with @joostjager and decided to not populat this field in `FetchAllInvoices()` but rather return to a previously implemented solution, where we select invoices and payment hashes in a differetn call: `FetchAllInvoicesWithPaymentHashes()`.

At one point it would be nice to change ChannelDB to also include the payment hash in Invoice  but it is not possible to do this in all cases without a larger refactor and DB migration in the current implementation.

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -505,48 +511,68 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+
+			// Save the invoice num as we'll need it at the and to restore order.
+			fetchedInvoiceNums = append(
+				fetchedInvoiceNums, byteOrder.Uint32(invoiceNum))
```
Close paren should being on new line below.

bhandras replied:
done

Roasbeef commented on clock/interface.go:
```diff
@@ -0,0 +1,11 @@
+
+// Clock is an interface that provides a time functions for LND packages.
+// This is useful during testing when a concrete time reference is needed.
+type Clock interface {
```
Couldn't a closure of `now func() time.Time` or a lighter interface _within_ the package that this is actually used have sufficed? I'm not sure such a bare interface warrants a new package, as the interace itself doesn't specify any behavior.

bhandras replied:
If there's a need for a "test clock" (stubbed out now) with state driven behavior then if now() is simply a closure, then state needs to be global (apart from trivial cases).
Clock can also be extended with other stubbed out functions like time.After() <- time.Time etc.

joostjager replied:
I like the part where you can also get `time.After` from here. We keep re-inventing this in unit tests.

bhandras replied:
I can add `time.After` now or later as needed.

joostjager replied:
I'd say only add it when we need it. I like the `Clock` interface. In the mpp pr, I also made a clock (that also has `tickAfter`)

joostjager replied:
Ok, now we have it. Nice

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+	return e.Expiry.Before(other.(*invoiceExpiry).Expiry)
+}
+
+// InvoiceExpiryWatcher handles automatic invoice cancellation of expried
```
`godoc` comments should either flow as a single paragraph, or use a new line to delimit new paragraphs. As is right now, each sentence reserves unnecessary deadspace at the end.

bhandras replied:
ptal

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+// will add that to the watching queue as well.
+// If any of the watched invoices expire, they'll be removed from the watching
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice.
+type InvoiceExpiryWatcher struct {
```
Does this struct need to be publicly exported? Similarly, looking above in `htlcswitch`, it seems instead `now` (w/e that is) can be passed in, with the registry itself taking care of initialization of this struct.

bhandras replied:
As InvoiceExpiryWatcher is a dependency of InvoiceRegistry, it needs to be public. Private initialization of dependencies afaik is not recommended as it's not leading to testable code. I think this is a pretty good article on the subject: https://testing.googleblog.com/2008/08/by-miko-hevery-so-you-decided-to.html

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice.
+type InvoiceExpiryWatcher struct {
+	sync.Mutex
+	started bool
```
This deviates from the customary `started/stopped` atomic variables we use elsewhere in the codebase.

bhandras replied:
I prefer using the mutex as that way either `Start()` or `Stop()` runs.

bhandras replied:
See also discussion below. I'm fine with changing this but believe it's better (see comments).

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+// PrefetchInvoices fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the watcher. This is useful to
+// prepopulate the watcher with past, but active invoices upon start.
+func (ew *InvoiceExpiryWatcher) PrefetchInvoices(cdb *channeldb.DB) {
```
Pointer members are usually some short prefix of the target struct.

Roasbeef replied:
Why not just have the invoices be passed in? This would save us from needing to depend on `channeldb.DB`, and also make testing easier (no need to make a test db instance).

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+	if err != nil {
+		log.Errorf("Error fetching invoices from the database: %v", err)
+	} else {
+		for k := 0; k < len(invoices); k++ {
```
No need for seeking beyond the current index, so the simpl `for := range` loop can be used here.

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+// Start starts the the subscription handler and the main loop. Start() can
+// only be called once.
+func (ew *InvoiceExpiryWatcher) Start(
+	cancelInvoice func(lntypes.Hash) error) error {
```
Any particular reason we pass the cancel closure here instead of in the main constructor? 

Following from this and the above comment, we can introduce a pretty slim interface here to handle fetching+cancelling the set of invoices.

bhandras replied:
As `InvoiceExpiryWatcher` is a dependency of `InvoiceRegistry` it is passed during construction and needs to exist before the registry.

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+
+// Stop cancels the invoice subscription and stops the expiry handler loop.
+func (ew *InvoiceExpiryWatcher) Stop() {
+	ew.Lock()
```
Haven't seen an instance in this diff, but my spidey senses tell me that a deadlock may be inadvertently introduced here if all callers don't mind proper mutex usage. We use atomic variables in every other new sub-system within the codebase to side step this, as things quickly exit if a shutdown has already been initiated.

bhandras replied:
The lock is only locking the Start/Stop procedure. Such that `Stop` will wait for `Start` if signaled during `Start`. Could you elaborate on how this may deadlock?

joostjager replied:
I think if we'd use the lock inside the main loop as well. Agreed that this looks a bit risky. It may be correct now, but a future dev may use this lock for something else too.

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+	log.Debugf("Adding invoice '%v' to expiry watcher, expiration: %v",
+		paymentHash, expiry)
+
+	ew.newInvoices <- &invoiceExpiry{
```
This should also select on the `quit` channel, otherwise a caller may lock if the watcher is concurrently shutting down.

bhandras replied:
done

Roasbeef commented on zpay32/invoice.go:
```diff
@@ -171,6 +171,16 @@ type Invoice struct {
+// ToInvoiceExpiry returns the passed expiry if it is non-zero, otherwise
+// it returns the default one hour.
+func ToInvoiceExpiry(expiry time.Duration) time.Duration {
+	if expiry == 0 {
```
What's this add? It also hard codes a _new_ default expiry rather than relying on the existing behavior to use the default expiry when decoding an invoice if unspecified.

bhandras replied:
Removed.

Roasbeef commented on invoices/invoice_expiry_watcher_test.go:
```diff
@@ -0,0 +1,46 @@
+)
+
+// Tests that invoices are canceled after expiration.
+func TestInvoiceExpiry(t *testing.T) {
```
Other cases that could be exercised (non-exhaustive list):
  * empty queue, no invoices cancelled
  * full queue, no invoices cancelled as not yet past expiry 
  * populated queue, add new invoice has already expired in the past 
  * populated queue, add new invoice, should be woken up to cancel

bhandras replied:
Will add a few more tests in an upcoming commit.

Roasbeef commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+// PrefetchInvoices fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the watcher. This is useful to
+// prepopulate the watcher with past, but active invoices upon start.
+func (ew *InvoiceExpiryWatcher) PrefetchInvoices(cdb *channeldb.DB) {
```
Why no return value? (the error)

bhandras replied:
Returns an error now. PTAL.

Roasbeef commented on queue/priority_queue.go:
```diff
@@ -0,0 +1,65 @@
+	Less(other PriorityQueueItem) bool
+}
+
+type priorityQueueRepr []PriorityQueueItem
```
What's "repr" here mean?

bhandras replied:
It meant "representation" but removed it to be less confusing.

Roasbeef commented on queue/priority_queue.go:
```diff
@@ -0,0 +1,65 @@
+func (eq *priorityQueueRepr) Pop() interface{} {
+	old := *eq
+	n := len(old)
+	item := old[n-1]
```
The final element should be set to nil in order to free it for collection.

bhandras replied:
Done.

Roasbeef commented on server.go:
```diff
@@ -387,7 +388,8 @@ func newServer(listenAddrs []net.Addr, chanDB *channeldb.DB,
 
 		invoices: invoices.NewRegistry(
-			chanDB, defaultFinalCltvRejectDelta,
+			chanDB, invoices.NewInvoiceExpiryWatcher(clock.DefaultClock{}),
```
See comment above about the registry just initializing this itself given a closure for the current time. Given that, I don't think this struct needs to be private.

bhandras replied:
Please see my above comment, but in general, dependencies should be injected rather privately constructed.

joostjager replied:
I also have a slight preference for passing in just the `Clock`.

Roasbeef commented on invoices/invoiceregistry.go:
```diff
@@ -396,6 +406,8 @@ func (i *InvoiceRegistry) AddInvoice(invoice *channeldb.Invoice,
 		return 0, err
 	}
 
+	i.expiryWatcher.AddInvoice(paymentHash, invoice)
```
What about checking the status of an invoice within `NotifyExitHopHtlc`? Otherwise, it's possible that concurrent with an invoice's expiry to wake up the main watcher loop, an invoice arrives (past expiration), which is then settled. Let's not lose sight of [_fundamental_ issue](https://github.com/lightningnetwork/lnd/issues/3448) we're trying to address here. 
>

joostjager replied:
To me, it seems that the worst that can happen is that you end up with a settled invoice in the database that has a `settle date` that is a fraction of a second after the expiry. It would indeed be better to make it so that this is never possible, but I I don't think it is important enough to address in this PR.

bhandras commented:
Decided to tidy up this pull request. All noise and past implementatons are hopefully removed now.

bhandras commented:
> I ended up reviewing the latest version of this diff as a whole since the commit structure appears to be in flux.
> 
> One thing which is missing from this current PR is updated integration tests to ensure relevant callers are updated once an invoice expires.

Please note that there's a "unit" test in invoiceregistry_test.go where it tests that the invoice is canceled and subscribers get a state update. I believe this is equivalent to the integration test.

bhandras commented:
Rebased on a fresh master and some commits are squashed to make it even more simple to review :)

cfromknecht commented on queue/priority_queue.go:
```diff
@@ -0,0 +1,66 @@
+	"container/heap"
+)
+
+type PriorityQueueItem interface {
```
missing godoc

cfromknecht commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -51,47 +37,55 @@ type InvoiceExpiryWatcher struct {
-	// subscription may not hold a valid payment hash.
-	netParams *chaincfg.Params
+	// quit signals InvoiceExpiryWatcher to stop.
+	quit chan interface{}
```
quit is reordered? also only needs to be `chan struct{}`

cfromknecht commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -123,26 +135,24 @@ func (ew *InvoiceExpiryWatcher) nextExpiry() time.Duration {
 			}
 			ew.expiryQueue.Pop()
+		} else {
+			break
```
could `break` in the negated conditional, in order to reduce the nesting

joostjager commented on channeldb/invoices.go:
```diff
@@ -546,6 +546,76 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+	err := d.View(func(tx *bbolt.Tx) error {
+		invoices := tx.Bucket(invoiceBucket)
+		if invoices == nil {
+			return ErrNoInvoicesCreated
```
Shouldn't it return an empty list in this case?

bhandras replied:
Hmm this is weird, looks like I lost at least one commit during rebase?

bhandras replied:
In my opinion it may be better to stay consistent with `FetchAllInvoices` and return nil and the error.

joostjager replied:
`invoiceBucket` is created when a new database is created, but that isn't true for `invoiceIndexBucket`. `invoiceIndexBucket` isn't accessed in `FetchAllInvoices`, so the problem won't show there. 

I think the code as it currently is will generate an error with an empty database?

bhandras replied:
Modified as discussed offline, and added a unit test as well.

bhandras replied:
It now only returns an error if the`invoiceBucket` does not exists (as that indicates a real error)

joostjager replied:
Returning nil happens in later commit, but should be done here.

cfromknecht commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -125,15 +124,15 @@ func (ew *InvoiceExpiryWatcher) AddInvoice(
-// expire. If there are no active invoices, then it simply returns max duration.
-func (ew *InvoiceExpiryWatcher) nextExpiry() time.Duration {
+// nextExpiry returns a Time chan to wait on until the next invoice expires.
+// If there are no active invoices, then it'll simply wait indefinitly.
```
seems like this could be done in the commit where it's introduced?

joostjager commented on channeldb/invoices.go:
```diff
@@ -546,6 +546,61 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+// only unsettled invoices and their payment hashes will be returned, skipping
+// all invoices that are fully settled.
+func (d *DB) FetchAllInvoicesWithPaymentHash(pendingOnly bool) (
+	[]Invoice, []lntypes.Hash, error) {
```
It seems that this code is changed later on in this pr, maybe because of a rebase mistake?

cfromknecht commented on invoices/invoiceregistry.go:
```diff
@@ -106,7 +106,10 @@ func NewRegistry(cdb *channeldb.DB, expiryWatcher *InvoiceExpiryWatcher,
 // Start starts the registry and all goroutines it needs to carry out its task.
 func (i *InvoiceRegistry) Start() error {
-	i.expiryWatcher.Start(i.CancelInvoice)
+	if err := i.expiryWatcher.Start(i.CancelInvoice); err != nil {
```
is this commit a workaround in case `InvoiceRegistry.Start` is called twice? why not add `sync.Once` pattern here and for Stop instead?

joostjager replied:
I think I suggested in an earlier revision to use the (atomic) `started`/`stopped` pattern. I think it is better to error when `Start` or `Stop` are called multiple times, to catch possible bugs.

bhandras replied:
Initially I used `sync.Once` but following reviewers comments decided to rework it. This intends to let either `Start` or `Stop` running, but let `Stop` wait for `Start` to finish if called "early". 

As for whether it's better to error:
While `Start` will error if called multiple times, imho it's better to let `Stop` be called multiple times in the in a start/stop situation as multiple objects may hold a reference to the instance that needs to be stopped, but this way only one will actually stop, all the others will fall trough (without error).

cfromknecht replied:
> to catch possible bugs.

what bugs are we preventing exactly? it's a singleton object, we don't care if it is started or stopped multiple times. the important thing is that any callers block until the operation has been completed

cfromknecht replied:
>  let Stop be called multiple times in the in a start/stop situation as multiple objects may hold a reference to the instance that needs to be stopped, but this way only one will actually stop, all the others will fall trough (without error).

I agree here, stopping should definitely support multiple calls without failure

joostjager replied:
I think in the way we use it, it should never be started multiple times. Maybe it is called from a different function that itself is really supposed to be called only once. In that case, an error here would expose that. All solutions work, so maybe it is mostly academical. I like the opportunity to catch those things.

cfromknecht replied:
i understand that we don't expect it to be called multiple times, but there is no harm in trying to start something that has already been started. so long as the resource is ready, the caller can begin using the api. it's incredibly unlikely that we will accidentally start the invoice twice in production, and even if we do, failing doesn't actually prevent any unintended consequences. all of this doesn't seem worth it to fix an benign, unlikely bug that only results in more code complexity and lower test coverage.

joostjager replied:
>it's incredibly unlikely that we will accidentally start the invoice twice in production

Still those things happen in software projects in general. all of a sudden there are two instances of the same thing.

What failing does is making us developers aware of a potential bug instead of silently ignoring an unexpected sequence of events.

To me, the atomic add or the sync.once isn't very different in terms of complexity.

But this is totally non-blocking. I am fine with sync.once too.

cfromknecht replied:
maybe it is just a difference in philosophy. my opinion is that if we have the ability to make the API robust against user error in a way which doesn't have negative consequences we should take that opportunity. the fact that a user can take the object in an unknown state, call start, and safely begin using is a feature, not bug imo :)

bhandras replied:
So to summarize `InvoiceExpiryWatcher.Start()` can be called only once, and will return an error otherwise (unless stopped), while `InvoiceExpiryWatcher.Stop()` can be called any times and the first one will stop the watcher.

I personally believe this is a better approach than the compare and exchange or the `sync.Once` solution  as it's more flexible but happy to revert this in favor of the one we use in the code if needed.

cfromknecht commented on channeldb/invoices.go:
```diff
@@ -546,28 +546,42 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+// payment hash. This struct is only used to store results of
+// ChannelDB.FetchAllInvoicesWithPaymentHash() call.
+type InvoiceWithPaymentHash struct {
+	Invoice     Invoice
```
exported fields need godocs

bhandras replied:
To me it seems this has a godoc. What am I missing?

cfromknecht replied:
the struct has one, but the fields do not

bhandras replied:
Sorry, missed that, fixed.

joostjager commented on channeldb/invoices.go:
```diff
@@ -546,6 +546,76 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+				Invoice: invoice,
+			}
+
+			copy(invoiceWithPaymentHash.PaymentHash[:], k[:lntypes.HashSize])
```
Why the `[:HashSize`]?

joostjager commented on htlcswitch/mock.go:
```diff
@@ -791,8 +792,8 @@ func newMockRegistry(minDelta uint32) *mockInvoiceRegistry {
 	finalCltvRejectDelta := int32(5)
-
-	registry := invoices.NewRegistry(cdb, finalCltvRejectDelta)
+	expiryWatcher := invoices.NewInvoiceExpiryWatcher(&clock.DefaultClock{})
```
`NewDefaultClock`?

cfromknecht commented on channeldb/invoices.go:
```diff
@@ -546,28 +546,42 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+// array is not ordered by add index.
 func (d *DB) FetchAllInvoicesWithPaymentHash(pendingOnly bool) (
-	[]Invoice, []lntypes.Hash, error) {
+	[]InvoiceWithPaymentHash, error) {
```
i wonder, is there reason this can't be done when FetchAllInvoicesWithPaymentHash is introduced? is there another dependency that prevents that?

bhandras replied:
Do you mean why we don't just return an optional `PaymentHash` field in the `Invoice`? I had an incarnation (lost in the numerous reviews) where this was done so, but seems like the introduced inconsistency that not all calls can populate the payment hash due to how the indexes are done wasn't worth it and reviewers requested to have a separate `FetchAllInvoicesWithPaymentHash` call instead. 

The reason why this is hard to do for all calls is because we have to iterate on the `invoiceIndex` as that maps from payment hash => invoice add idx, while the `invoiceBucket` maps from invoice add idx => invoice. So only calls where we refer to invoices by their payment hash can be changed to populate payment hash.

cfromknecht commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -68,22 +66,29 @@ func NewInvoiceExpiryWatcher(clock clock.Clock) *InvoiceExpiryWatcher {
+		return err
+	}
+
+	for k := 0; k < len(pendingInvoices); k++ {
```
`for i := range pendingInvoices {`?

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,198 @@
+	ew.Lock()
+	defer ew.Unlock()
+
+	if ew.started {
```
A pattern that we use elsewhere is an atomic add: `if atomic.AddInt32(&n.started, 1) != 1 {`

Then the lock isn't needed.

bhandras replied:
I understand that, but to me that pattern fails when for some reason (early stopping) `Stop()` is called during `Start()`. By locking `Stop()` will wait and do the stopping, with compare exchange it'll simply return.

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,198 @@
+
+	expiry := invoice.CreationDate.Add(
+		zpay32.ToInvoiceExpiry(invoice.Terms.Expiry),
+	)
```
This looks a bit indirect to me. Why not just expose the default from `zpay32` and do an if statement here?

bhandras replied:
Initially I think that was the case but then due to review comments I made that function in zpay32. Now reverted as you requested.

bhandras commented:
> Completed an initial pass, most of the comments are just nits or commit structure questions
> 
> Biggest thing i see right now is that it's unclear what happens during a restart. It seems like we don't repopulate the expiry watcher with preexisting invoices, and that it is only watching new invoices sent over the invoice subscription. Is that the case?

Thanks for the review @cfromknecht ! :)
There's a prefetch happening in `InvoiceRegistry.Start()` so it'll fetch all pending invoices.

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -96,21 +99,33 @@ func NewRegistry(cdb *channeldb.DB, finalCltvRejectDelta int32) *InvoiceRegistry
 
+	if err := i.expiryWatcher.PrefetchInvoices(i.cdb); err != nil {
+		return err
+	}
```
Could merge these calls, prefetch inside start. Unless we use them already anywhere separately.

bhandras replied:
Initially I wanted the API to be more flexible, but now that it's finished I agree with you, there's no need to separate this. Done.

bhandras replied:
So prefetch is now part of `InvoiceRegisty`. This way `InvoiceExpiryWatcher` won't depend on `channeldb.DB` (suggested by @Roasbeef)

joostjager commented on invoices/invoiceregistry_test.go:
```diff
@@ -654,18 +542,77 @@ func TestUnknownInvoice(t *testing.T) {
+		}
+	case <-time.After(testTimeout):
+		t.Fatal("no update received")
+	}
```
May I shill `defer timeout(t)()` to you as well? I personally don't like the bloat of selecting on timeout channels.

bhandras replied:
This test is now updated. I'd love to use `defer timeout(t)()` but we need to actually wait here before the expectation checks.

joostjager commented on invoices/invoiceregistry_test.go:
```diff
@@ -654,18 +542,77 @@ func TestUnknownInvoice(t *testing.T) {
+	// Add the invoice and expect that AddInvoice is successful.
+	addIdx, err := ctx.registry.AddInvoice(
+		expiringInvoice, testInvoicePaymentHash,
+	)
```
I would at least test the normal flow here as well where an invoice is not yet expired.

bhandras replied:
After some thinking I decided to test expiry but not not through the subscribing for invoice state changes as it's quiet difficult due to all invoices requiring their own subscriptions. It is also problematic as with subscriptions as there's a possibility that state change remains unobserved when subscribing after `InvoiceRegisty.Start()` (cancellation might happen before the subscription is active).

I wonder if there's any simpler but also more correct ways to test this.

joostjager replied:
What I mean is that I miss manipulation the test clock to let invoices expire.

cfromknecht commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,215 @@
+
+// ExpiryWatcherClock is used to provide a time functions for ExpiryWatcher,
+// which can be stubbed out in tests.
+type ExpiryWatcherClock interface {
```
another thought, since the `clock` package is primarily for testing we can expose a `MockClock` that way it doesn't have to be redefined everywhere it is used

bhandras commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,215 @@
+				invoice.State != channeldb.ContractSettled {
+				// Try to add the invoice to the expiry watcher queue.
+				if err := ew.addInvoice(invoice); err != nil {
+					log.Errorf("Error adding invoice to the expiry watcher queue: %v", err)
```
done

bhandras commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,215 @@
+		// Subscribe to all invoice notificatons and start a goroutine to
+		// handle the subscription and forward all new invoices to the expiry
+		// queue.
+		ew.subscription = ew.registry.SubscribeNotifications(0, 0)
```
Sorry, this is a rebase issue, you shouldn't even see this code. I'll try to fix this.

bhandras commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,215 @@
+	// in the invoice.
+	paymentHash, err := ew.decodePaymentHash(invoice)
+	if err != nil {
+		return fmt.Errorf("Could not decode payment hash for invoice: %v", invoice)
```
Again this code should not be here, it's an outdated version and you only see this because of a rebase error on my side. I'll try to fix the commit structure again.

bhandras commented on invoices/test_helpers.go:
```diff
@@ -153,6 +153,7 @@ func newTestInvoice(t *testing.T,
 			PaymentPreimage: testInvoicePreimage,
 			Value:           testInvoiceAmount,
 			Expiry:          expiry,
+			Features:        testFeatures,
```
fixed

bhandras commented:
Squashed again to remove deleted history bleeding into the reviews. Hopefully this 4 commits provides a much cleaner review experience.

joostjager commented on queue/priority_queue.go:
```diff
@@ -0,0 +1,66 @@
+type priorityQueue []PriorityQueueItem
+
+// Len returns the length of the ExpiryQueue.
+func (pq priorityQueue) Len() int { return len(pq) }
```
There were some comments on this commit that I cherry-picked in https://github.com/lightningnetwork/lnd/pull/3415. You may want to replace this commit by the updated version.

bhandras replied:
Fixed the godocs from your commit.

joostjager commented on clock/interface.go:
```diff
@@ -0,0 +1,11 @@
+// Clock is an interface that provides a time functions for LND packages.
+// This is useful during testing when a concrete time reference is needed.
+type Clock interface {
+	Now() time.Time
```
Document exported method

bhandras replied:
done

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// Less implements PriorityQueueItem.Less such that the top item in the
+// priorty queue will be the one with that expires next.
```
language: "one with that expires next"

joostjager commented on invoices/invoice_expiry_watcher.go:
```diff
@@ -0,0 +1,193 @@
+			top.PaymentHash, top.Expiry)
+
+		err := ew.cancelInvoice(top.PaymentHash)
+		if err != nil && err != channeldb.ErrInvoiceAlreadySettled &&
```
Does this potentially cancel an invoice in the `Accepted` state? That shouldn't happen.

bhandras replied:
Thanks for pointing this out. As discussed offline, this is now handled in `InvoiceRegistry.cancelInvoiceImpl`.

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -95,21 +99,51 @@ func NewRegistry(cdb *channeldb.DB, finalCltvRejectDelta int32) *InvoiceRegistry
 
-	return nil
+	// Now prefetch all pending invoices to the expiry watcher.
+	return i.prefetchPendingInvoices()
```
Is this the right order? Start go routine before prefetch?

bhandras replied:
Yes, `InvoiceExpiryWatcher` must be started before `InvoiceExpiryWatcher.AddInvoice` may be called, otherwise it'll block until it is not started.

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -95,21 +99,51 @@ func NewRegistry(cdb *channeldb.DB, finalCltvRejectDelta int32) *InvoiceRegistry
 
+// prefetchPending fetches all active invoices and their corresponding payment
+// hashes from ChannelDB and adds them to the expiry watcher.
+func (i *InvoiceRegistry) prefetchPendingInvoices() error {
```
Function name doesn't fully cover what it does. It doesn't just fetch, but also primes the watcher.

bhandras replied:
Good point, renamed it to better express what it does.

joostjager commented on invoices/invoiceregistry_test.go:
```diff
@@ -547,3 +550,84 @@ func TestUnknownInvoice(t *testing.T) {
+	// First prefill the Channel DB with some pre-existing invoices,
+	// half of them still pending, half of them expired.
+	const numExpired = 50
+	const numPending = 50
```
Does it add something to use that many test invoices?

bhandras replied:
Changed it. PTAL

joostjager commented on invoices/invoiceregistry_test.go:
```diff
@@ -547,3 +550,84 @@ func TestUnknownInvoice(t *testing.T) {
+		}
+	}
+
+	<-time.After(testTimeout)
```
`time.Sleep`?

joostjager replied:
I would expect the test clock being manipulated here. That is what it is for right?

bhandras replied:
Thanks, not sure why I use `time.After` afterall :)
In this case we must wait (time is already manipulated in the test), as `InvoiceExpiryWathcer` must go trough all invoices and cancel them through `InvoiceRegistry`. We may skip the waiting by waiting for state changes on each invoice's invoice subscriptions. I decided to instead use `LookupInvoice` (as discussed with you offline).

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -610,6 +669,13 @@ func (i *InvoiceRegistry) CancelInvoice(payHash lntypes.Hash) error {
 
+	// Return without cancellation if the invoice state is ContractAccepted.
+	if invoice.State == channeldb.ContractAccepted {
+		log.Debugf("Invoice(%v): is remains accepted as cancel wasn't"+
```
Check sentence

bhandras replied:
Thanks, done.

bhandras commented on invoices/invoice_expiry_watcher_test.go:
```diff
@@ -0,0 +1,124 @@
+	return test
+}
+
+func (test *invoiceExpiryWatcherTest) checkExpectations() {
```
Corrected to conform this. (I personally like to use longer names for readability and reserve single character variable names for loop indexes or math code.)

bhandras commented on invoices/test_helpers.go:
```diff
@@ -193,3 +210,45 @@ func newTestInvoice(t *testing.T,
+		var preimage lntypes.Preimage
+		if _, err := crand.Read(preimage[:]); err != nil {
+			t.Fatalf("can't read crypto random: %v", err)
+		}
```
Initially I used deterministic preimages but decided to rewrite it to use crypto random instead as many tests seem to rely on that. Redone with deterministic as suggested.

bhandras commented on invoices/invoiceregistry.go:
```diff
@@ -95,21 +99,54 @@ func NewRegistry(cdb *channeldb.DB, finalCltvRejectDelta int32) *InvoiceRegistry
+	if err := i.expiryWatcher.Start(func(paymentHash lntypes.Hash) error {
+		cancelIfAccepted := false
+		return i.cancelInvoiceImpl(paymentHash, cancelIfAccepted)
+	}); err != nil {
```
Done.

bhandras commented on invoices/invoiceregistry.go:
```diff
@@ -95,21 +99,54 @@ func NewRegistry(cdb *channeldb.DB, finalCltvRejectDelta int32) *InvoiceRegistry
 
 // Stop signals the registry for a graceful shutdown.
 func (i *InvoiceRegistry) Stop() {
+	i.expiryWatcher.Stop()
```
I think it's not possible as in that case `ExpiryWatcher` may cancel invoices even though the the notification handler has quit. It'd probably block writing to the notification handlers channel.

joostjager replied:
`notifyClients` checks the registry's `quit` channel. But you can also leave it as is. Both objects are depending on each other.

joostjager commented:
LGTM. Holding off further review until second reviewer has left a review.

joostjager commented:
Needs rebase

cfromknecht commented on invoices/test_helpers.go:
```diff
@@ -0,0 +1,195 @@
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+type mockPayload struct {
```
these structs are now intermingled w/ the package instead only being compiled with the tests, is that intentional? not a blocker

bhandras replied:
Thanks for pointing out, I somehow forgot Go does this. Renamed `test_helper.go` to `helper_test.go` to avoid this.

cfromknecht commented on queue/priority_queue.go:
```diff
@@ -0,0 +1,76 @@
+type PriorityQueueItem interface {
+	// Less must return true if this item is ordered before other and false
+	// otherwise.
+	Less(other PriorityQueueItem) bool
```
nice work on this interface!

bhandras replied:
Thanks :) It's now merged through an other PR, but fixed the comments in my commit to be fully equal.

joostjager replied:
I really thought I took that comment on `Less`. Ok, next time we should split off the common functionality between two prs.

cfromknecht commented on queue/priority_queue_test.go:
```diff
@@ -0,0 +1,67 @@
+	// The number of elements we push to the queue.
+	count := 100
+	// Generate a random permutation of a range [0, count)
+	array := rand.Perm(count)
```
🤓

bhandras replied:
:cowboy_hat_face:

cfromknecht commented:
will need a rebase tho

bhandras commented:
Rebased on top of the latest master. As discussed offline I made some changes in the `clock.Clock` intro commit to make InvoiceRegistry use that interface too and move the private `invoices.testClock` to `clock.TestClock` and added basic tests. PTAL.

joostjager commented on clock/log.go:
```diff
@@ -0,0 +1,26 @@
+// using btclog.
+func UseLogger(logger btclog.Logger) {
+	log = logger
+}
```
Log is unused? Also seems a bit too much maybe, to log for `clock`.

bhandras replied:
I added the log because there's a debug log in `TestClock` (you added that actually). Would you like to remove it? (I too think logging is not necessary).

joostjager replied:
Ah, that is right. Yes, I added the logging because it was quite useful to debug the tests. Maybe we could otherwise pass in a logger to `TestClock`? Or you can remove it if you don't like that.

As a side note, if you add a typical `log.go` file, it also needs to be connect in the root `log.go`.

bhandras replied:
I'm removing it then as there's no need to debug (tests added).

joostjager replied:
Ok. To clarify: I didn't add them to debug the test clock, but to debug the test that was using the test clock.

joostjager commented on channeldb/invoices.go:
```diff
@@ -594,7 +594,11 @@ func (d *DB) FetchAllInvoicesWithPaymentHash(pendingOnly bool) (
+			// index as that simply means there are no
+			// invoices added yet to the DB. In this case
+			// we simply return an empty list.
+			return nil
```
Ah, this needs to go to the earlier commit. Maybe you can group the addition of `FetchAllInvoicesWithPaymentHash` itself with this test.

bhandras replied:
Sorry, forgot to move this the the earlier one. Fixed.

joostjager commented on invoices/helpers_test.go:
```diff
@@ -0,0 +1,209 @@
+package invoices
```
I think we generally call this file `utils_test.go`

bhandras replied:
Moved to `test_utils_test.go` as discussed offline.

joostjager commented on invoices/helpers_test.go:
```diff
@@ -0,0 +1,209 @@
+package invoices
```
A separate commit that does a pure move makes it easier to review the addition. But no need to change now.

bhandras replied:
Can separate it if you really want to...

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -145,21 +149,56 @@ func NewRegistry(cdb *channeldb.DB, cfg *RegistryConfig) *InvoiceRegistry {
+	// Start InvoiceExpiryWatcher and prepopulate it with existing active
+	// invoices.
+	err := i.expiryWatcher.Start(func(paymentHash lntypes.Hash) error {
+		cancelIfAccepted := false
```
I am not sure about this pattern. It works, it is self documenting, but if we'd pass in every constant bool like this, not sure if it is generally liked.

bhandras replied:
Imho it's much more readable and don't really see any issues of having these here and there.

joostjager commented on invoices/invoiceregistry.go:
```diff
@@ -145,21 +149,56 @@ func NewRegistry(cdb *channeldb.DB, cfg *RegistryConfig) *InvoiceRegistry {
 
-	return nil
+	// Now prefetch all pending invoices to the expiry watcher.
+	return i.populateExpiryWatcher()
```
If this returns an error, where do we stop the goroutine?

bhandras replied:
Fixed.

bhandras commented:
Re: manipulating time in invoiceregistry test: fixed that.

joostjager commented on channeldb/invoices.go:
```diff
@@ -565,6 +565,83 @@ func (d *DB) LookupInvoice(paymentHash [32]byte) (Invoice, error) {
+		invoiceIndex := invoices.Bucket(invoiceIndexBucket)
+		if invoiceIndex == nil {
+			// Mask the error if there's an invoice
+			// index as that simply means there are no
```
"if there is _no_ invoice index".

bhandras replied:
Done.

joostjager commented on invoices/invoiceregistry_test.go:
```diff
@@ -636,3 +640,87 @@ func TestSettleMpp(t *testing.T) {
+	}
+
+	// Fwd time 1 day.
+	testClock.SetTime(testTime.Add(24 * time.Hour))
```
Ideally you'd also want to assert that they were not canceled yet before the clock was advanced. (non blocking)

bhandras replied:
That's a good idea, added it.
